Acceptance tests (testing CI)

ekinanp opened the pull request.

ekinanp commented on manifests/trusted_external_command.pp:
```diff
@@ -69,7 +69,10 @@
   ])
 
-  unless Service['pe-puppetserver'] {
+  unless defined(Service['pe-puppetserver']) {
```
@reidmv Do you think we could remove the `defined` check and just declare the `puppetserver` resource? I remember you mentioning that doing that could lead to a slight performance hit, so wanted to double check.

reidmv replied:
This stanza should be fully removed, and the `pe-puppetserver` resource should **not** be declared.

The `defined()` function can reliably determine if a class definition is available (whether or not it exists and could be included), but `defined()` is unreliable for determining if a resource _will be_ present in the catalog—it is evaluation-order dependent. If the `puppet_enterprise::master` class has been evaluated first, then this `defined()` would return `true`. If the `puppet_enterprise::master` class has not been evaluated _yet_, this `defined()` will return false, the resource will be declared, and when `puppet_enterprise::master` is evaluated later, a duplicate-resource-definition error will be raised.

This class should depend on and require that something else declare `Service[pe-puppetserver]`. It should never declare that service itself.

There are other ways this can be addressed too but IMO the extra complexity and parameterization (or resource collector tricks) do not sufficiently improve the UX to be worth the complexity cost.

ekinanp replied:
Thanks for the feedback. Will remove.

ekinanp commented on spec/classes/trusted_external_command_spec.rb:
```diff
@@ -23,14 +23,7 @@
-  # context "when Service['pe-puppetserver'] is undefined" do
-  #   it { is_expected.to compile }
-  # end
+  context "when Service['pe-puppetserver'] is undefined" do
```
I use `puppet apply` when applying the `trusted_external_command` class so I think this test is useful to make sure that use-case works.

RandomNoun7 replied:
It's not that the test isn't useful, it's that it doesn't work properly if a puppet run is called via `puppet agent -t`. Using puppet apply will work fine to verify that the class can configure a master properly, but using puppet apply does not trigger the external command to actually be called. To trigger to external command to be called and return data so that things like `$trusted` data are populated, an agent run must be triggered via puppet agent -t. Then what you find is that if you use `defined` as the guard for `Service['pe-puppetserver]` resource the run fails because it doesn't properly protect against duplicate resource declarations like you would hope it would. So, to get a real puppet agent -t run to work you have to remove `defined` and use something else that really will guard against duplicate resource declarations. Unfortunately I haven't yet found a syntax to use in the manifest that will both fix puppet runs and work with this test.

RandomNoun7 replied:
Based on the feedback from @reidmv aboce this test would be removed anyway because the assumption is that this resource would always be declared by some other file.

ekinanp replied:
Yup, just saw his feedback. I'll remove the test.

ekinanp commented on .fixtures.yml:
```diff
@@ -20,5 +20,3 @@ fixtures:
     facts: 'git://github.com/puppetlabs/puppetlabs-facts.git'
     puppet_agent: 'git://github.com/puppetlabs/puppetlabs-puppet_agent.git'
     provision: 'git://github.com/puppetlabs/provision.git'
-  symlinks:
```
@RandomNoun7 Is this OK? Wasn't sure why this was added.

RandomNoun7 replied:
As long as it creates the proper symlink in `spec/fixtures/modules` when `spec_prep` is called, sure. It was probably required at one point and maybe is no longer.

ekinanp replied:
Yeah, things were working OK without it so looks like it is no longer needed.

ekinanp commented on spec/acceptance/trusted_external_command_spec.rb:
```diff
@@ -14,48 +14,24 @@
-      expect(mockserver.mock_called?(default_endpoint, default_query_params, 20, 40)).to be true
-    end
-
-    it 'applies the correct yaml settings' do
```
@RandomNoun7 I think we should test "creates correct config file" in the unit tests. We can do that by checking that the `trusted_external_command` resource creates the relevant `File` resource with the right YAML content.

RandomNoun7 replied:
Testing in unit tests is fine, but it should also be tested in acceptance that the file is created and the content is as expected.

ekinanp replied:
Why? Aren't we already testing that when checking that the script works? If the config file wasn't created w/ the right content, how can the script work in the first place? That's what I meant by implementation detail.

RandomNoun7 replied:
If you feel strongly that it's not needed I'm fine with that. I'm usually of the attitude that more tests is always better and testing the data storage layer is usually a good idea.

ekinanp commented on spec/unit/files/servicenow_spec.rb:
```diff
@@ -30,8 +30,8 @@
 
   context 'node exists' do
     it "returns the node's parsed CMDB record" do
```
@RandomNoun7 Do you agree w/ my rationale in the commit message re: this change?

RandomNoun7 replied:
Sorry, but I don't think I do. The expected result of running the external node data script to query a servicenow instance is not to output the raw cmdb record. The data output by the script will be munged in a few different ways that make the expectation that the node data hash output by the script should match the cmdb record returned by servicenow not quite correct.

ekinanp replied:
We're thinking of "trusted external data" and "classification" as separate features. The "trusted external data" feature should output just the raw CMDB record. The only munging that the script does is for the "classification" bit and in that bit, we're already testing all of that munging.

This test is testing the "trusted external data" bit (when there is no "classification"). That bit is supposed to output the raw CMDB record.

RandomNoun7 replied:
Ok, I think I finally see the different tests you're trying to run here. When I initially ran these acceptance tests with a realistic api response json as the mock this test fell apart and I think it confused me.

ekinanp commented on spec/acceptance/trusted_external_data_spec.rb:
```diff
@@ -0,0 +1,72 @@
+require 'spec_helper_acceptance'
+
+context 'trusted external data' do
```
I think acceptance tests should be structured as "We're testing this feature." In this case, the feature's "trusted external data". "trusted external command" is an implementation detail.

ekinanp commented on spec/acceptance/trusted_external_data_spec.rb:
```diff
@@ -0,0 +1,72 @@
+context 'trusted external data' do
+  # Set up the trusted external command on the master
+  #
+  # TODO: Should this move to suite-level or even Rakefile-level
```
I think yes. It should move to suite-level. Maybe even to the Setup rake tasks

RandomNoun7 replied:
Moving to suite level seems fine, but I don't think rake is the right place. If the setup fails for some reason I want rspec to tell me about it and fail the run via a test failure of some sort.

ekinanp replied:
Yeah that's fine. I'll move it to suite-level then.

ekinanp commented on Rakefile:
```diff
@@ -86,3 +86,211 @@ EOM
+  task :ci_run_tests do
+    begin
+      Rake::Task['acceptance:setup'].invoke
+      Rake::Task['acceptance:run_tests'].invoke 
```
Unfortunately I deleted my `inventory.yaml` file and now vmpooler's failing to get the right Ubuntu VM (on the API end, not us) so I'll post a high-level 'final' `inventory.yaml` after setup runs:
```
- groups:
      - name: ssh_nodes
         targets:
           - <master_vm_bolt_config>
      - name: servicenow_nodes
         targets:
           - <servicenow_instance_bolt_config> # Based on v2 inventory in https://github.com/puppetlabs/servicenow_tasks#setup
vars:
  cmdb_record_ids:
    <master_uri>: <master_cmdb_record_id>
```

I'll update this once vmpooler's working again.

RandomNoun7 replied:
Oops, I accidentally resolved a conversation I wasn't involved in. Sorry.

ekinanp commented on spec/support/acceptance/litmus_helpers.rb:
```diff
@@ -0,0 +1,74 @@
+  end
+end
+
+module LitmusHelpers
```
Rationale for this module is that some of Litmus' methods (e.g. `inventory_hash_from_inventory_file`) should be class-level. So, I think `LitmusHelpers.inventory_hash_from_inventory_file` is more readable than `LitmusHelpers.instance.<...>`.

RandomNoun7 replied:
Wouldn't a PR to litmus be a better place for this change? It seems like we should use Litmus'  available syntax and if we think that syntax is deficient we should fix Litmus use the new syntax when it becomes available.

ekinanp replied:
Yeah, there's some stuff I'd like to fix in Litmus. Problem is "put up PR" => "get changes reviewed" => "get new Litmus release" will take too long so would prefer to keep this for readability. I can add a "TODO" saying "Remove this once Litmus includes these as class methods".

ekinanp commented on spec/support/acceptance/litmus_helpers.rb:
```diff
@@ -0,0 +1,74 @@
+# These helpers are a useful way for tests to reuse
+# Litmus' helpers when they want to do stuff on nodes
+# that may not be the current target host (like e.g.
+# the master or the ServiceNow instance).
```
Example of when master isn't target host is when we have an agent, and the agent attempts to modify master's site.pp to do stuff.

ekinanp commented on spec/acceptance/trusted_external_data_spec.rb:
```diff
@@ -0,0 +1,72 @@
+  message => "#{separator}${trusted_json}#{separator}"
+}
+      MANIFEST
+      # TODO: This will cause some problems if we run the tests
```
This is probably an issue with Litmus' design.

ekinanp commented on spec/support/acceptance/servicenow/mock_instance.rb:
```diff
@@ -0,0 +1,107 @@
+# Our mock ServiceNow instance is a Sinatra server that mimics
+# the relevant ServiceNow API endpoints used by the tests
+#
+# TODO: See if we can get a persistent dev instance that only CI
```
Eh, I guess the mock instance makes stray-CMDB-entry cleanup easier. Also easier if there are concurrent tests running. So this TODO's probably unnecessary.

ekinanp commented:
The tests work locally, but they're failing CI b/c of vmpooler, and will probably fail once vmpooler's able to get us VMs since GH actions doesn't know how to connect to a vmpooler VM.

ekinanp commented on .fixtures.yml:
```diff
@@ -20,3 +20,4 @@ fixtures:
     facts: 'git://github.com/puppetlabs/puppetlabs-facts.git'
     puppet_agent: 'git://github.com/puppetlabs/puppetlabs-puppet_agent.git'
     provision: 'git://github.com/puppetlabs/provision.git'
+    servicenow_tasks: 'git://github.com/puppetlabs/servicenow_tasks.git'
```
@gsparks We should probably tag this. I guess I could also just point to a specific commit.

ekinanp commented:
Oh, I'll update the README once we've fleshed out a reasonable acceptance test UX.

ekinanp commented on spec/support/acceptance/litmus_helpers.rb:
```diff
@@ -0,0 +1,74 @@
+  old_target_host = ENV['TARGET_HOST']
+  target_obj = @targets[name]
+  begin
+    ENV['TARGET_HOST'] = target_obj.uri
```
The Litmus helpers assume `TARGET_HOST` is set, so this is why we are doing this.

ekinanp commented:
Next steps after this is to add a test for ServiceNow classification.

ekinanp commented on Rakefile:
```diff
@@ -86,3 +86,211 @@ EOM
+  end
+
+  desc 'Sets up the ServiceNow instance'
+  task :setup_servicenow_instance do
```
This task could also take a `config` hash containing dev instance config and just set-up the `inventory.yaml` file.

RandomNoun7 commented on spec/acceptance/trusted_external_data_spec.rb:
```diff
@@ -0,0 +1,72 @@
+    separator = '<trusted_json>'
+
+    before(:all) do
+      manifest = <<-MANIFEST
```
Could you please indent this at the same level as the `<<-` characters? Indenting all the way to the left of the file makes the file slightly harder to read.

ekinanp replied:
Yeah, I can do that.

RandomNoun7 commented on spec/acceptance/trusted_external_data_spec.rb:
```diff
@@ -0,0 +1,72 @@
+      puppet_output = result.stdout
+
+      trusted_json = nil
+      begin
```
This block, and the process of getting the `cmdb_record_id` are a lot of code to put in an `it` block. Could these be extracted to a helper file so the test reads more naturally?

ekinanp replied:
Yeah, it does look like a lot of code and the Rake tasks are already getting the `cmdb_record_id`. I can do that.

RandomNoun7 commented on spec/acceptance/trusted_external_data_spec.rb:
```diff
@@ -0,0 +1,72 @@
+      inventory_hash = LitmusHelpers.inventory_hash_from_inventory_file
+      cmdb_record_id = inventory_hash['vars']['cmdb_record_ids'][LitmusHelpers.instance.uri]
+
+      # The unit tests already check against the entire CMDB record JSON so it's
```
Just because unit tests are checking for something doesn't mean acceptance tests should avoid checking that thing. Unit tests are great at running fast and ensuring your assumptions about how code should work and interact with a system are running as expected, but acceptance tests are there to catch you and have your back in case any of those assumptions turn out to not hold true forever. I think at a minimum there should be tests to ensure that the fields we manipulate to change their name, and the fields we add to the record should also be tested.

ekinanp replied:
Acceptance tests should also be higher-level and test that the module is doing what we are claiming to do to the user. But I do agree that this assertion could be more thorough and doing so isn't too hard. I think I have an idea how to do that.

Re: "fields we manipulate to change their name/fields we add", all of that's for classification. We don't need to test those fields here -- this test file just tests the "trusted external data" feature.

RandomNoun7 commented on Rakefile:
```diff
@@ -86,3 +86,211 @@ EOM
+  end
+
+  desc 'Teardown the setup'
+  task :tear_down do
```
Is this really necessary? The mock server and the master are both running on the same pooler vm aren't they? Just call litmus' tear down task and it will throw away the vm. Why go to all this trouble to clean up a vm that's about to thrown away?

ekinanp replied:
This also clears out test-CMDB records if I end up using something other than the mock ServiceNow instance (like my dev instance). Notice that that code noops if the user ends up using the mock ServiceNow instance. So, I would like to keep it.

ekinanp commented on spec/spec_helper_acceptance_local.rb:
```diff
@@ -1,101 +1,64 @@
-  parsed_data.dig :external, :servicenow
+TRUSTED_JSON_SEPARATOR = '<TRUSTED_JSON>'.freeze
+def parse_trusted_json(puppet_output)
+  trusted_json = puppet_output.split(TRUSTED_JSON_SEPARATOR)[1]
```
Forgot to explain why I changed this to use the separator. Reason is b/c it felt more accurate than https://github.com/puppetlabs/puppetlabs-servicenow_cmdb_integration/blob/master/spec/spec_helper_acceptance_local.rb#L95, which could potentially fail if (for some reason) Puppet decides to change the `defined <...>` part of the output.

ekinanp commented on Rakefile:
```diff
@@ -86,3 +86,190 @@ EOM
+    master_uri = on_master(&:uri)
+    inventory_hash = inventory_hash_from_inventory_file
+    inventory_hash['vars'] = {
+      'cmdb_record_ids' => CMDBHelpers.record_ids.merge(master_uri => sys_id),
```
This shouldn't be necessary once `servicenow_tasks` has a `get_records` task b/c I can just query for CMDB entry where `fqdn == uri`.

ekinanp commented on spec/support/acceptance/helpers.rb:
```diff
@@ -0,0 +1,131 @@
+  end
+  module_function :on_servicenow_instance
+
+  def on_target(name, setup_task, classifying_var)
```
Per the convo with David, this `on_<target>` stuff should probably just be `<target>`, thus avoiding the use of the block.

ekinanp replied:
Helper should also take in a `role`, so each target should have a `var['roles']` that we check.

ekinanp commented on Rakefile:
```diff
@@ -86,3 +86,190 @@ EOM
+    # including our ServiceNow instance. Until litmus:acceptance:parallel
+    # has a way for you to pass-in specific targets to run the tests on,
+    # we'll just run these tests via rspec
+    system({ 'TARGET_HOST' => on_master(&:uri) }, 'bundle exec rspec ./spec/acceptance')
```
Also per the convo with David, we should probably just run `rspec` here. The `TARGET_HOST` doesn't make sense for our multi-node test environment.

RandomNoun7 replied:
Is a regular call to `system` here going to return all of the output streams properly? Or would it be better to call Open3 so you can examine stderr vs stdout and return the right thing?

ekinanp replied:
It outputs the right thing at the right streams. But it looks like `system` returns true if exit code was 0, false otherwise so this current code will exit 0 even if rspec fails. Will fix that.

ekinanp replied:
Updated.

ekinanp commented on spec/support/acceptance/helpers.rb:
```diff
@@ -0,0 +1,131 @@
+module LitmusHelpers
+  extend PuppetLitmus
+
+  def instance
```
This should be removed.

ekinanp commented:
@RandomNoun7 I updated the PR. The updates:

* Address David S' feedback. So all targets are still stored in `inventory.yaml` and each target has a role. Here's an example of what the `inventory.yaml` file looks like now:

```
---
version: 2
groups:
- name: docker_nodes
  targets: []
- name: ssh_nodes
  targets:
  - uri: okay-rattle.delivery.puppetlabs.net
    config:
      transport: ssh
      ssh:
        user: HIDDEN
        password: HIDDEN
        host-key-check: false
    facts:
      provisioner: vmpooler
      platform: ubuntu-1604-x86_64
    vars:
      roles:
      - master
- name: winrm_nodes
  targets: []
- name: servicenow_nodes
  targets:
  - uri: okay-rattle.delivery.puppetlabs.net:1080
    config:
      transport: remote
      remote:
        user: HIDDEN
        password: HIDDEN
    vars:
      roles:
      - servicenow_instance
```

I've also removed `LitmusHelpers.instance` and changed the `run_tests` task to invoke `rspec spec/acceptance` _without_ setting `TARGET_HOST`.

* The CMDB helpers have been changed from `(get|delete)_record` => `(get|delete)_target_record` thanks to the `get_records` task added earlier today. That considerably simplifies all the CMDB setup and tear down.

ekinanp commented on spec/support/acceptance/helpers.rb:
```diff
@@ -0,0 +1,116 @@
+  # will work as expected. Otherwise if we do this in, say, the
+  # constructor, then the code will only work for the agent that
+  # most recently set the TARGET_HOST variable.
+  PuppetLitmus.instance_methods.each do |name|
```
We probably won't need something like this, but this _is_ the right way to set `TARGET_HOST` and it isn't too hard to understand so I'd like to keep the general solution.
